### PR TITLE
[streaming/xcode] configurable MP3 streaming bitrate

### PR DESCRIPTION
**Pull request by whatdoineed2do**
MP3 streaming is set to 44.1k/16/2 with a bit rate of 128kbps.

Streaming bandwidth on an internal (home) networks are likely to be be able to support higher bitrates than 128kbps and it is desirable to have to ability to adjust upwards the bitrate, particularly if the endpoint feeds into a DAC.

This PR adds the `library.streaming_bitrate` configuration value to allow users to select 128/192/320 bitrates.

Closes https://github.com/ejurgensen/forked-daapd/issues/503

We extend `struct media_quality` to incl `bit_rate` which is normally `0` unless we require a specific bitrate for the transcode.  Currently the transcode's `AVCodecContext::bit_rate` is defaulted to 128000 and there is no impact on setting this to 0 since it looks like `bit_rate` field is ignored during the encode side (prep'ing the write to generate PCM)

```
$ ffmpeg -hide_banner -i http://localhost:3689/stream.mp3 -f pulse "output stream"
Input #0, mp3, from 'http://localhost:3689/stream.mp3':
  Metadata:
    icy-name        : My Music on rz4UOJ9JspE
    StreamTitle     : 太阳 - PikA 好声音
  Duration: N/A, start: 0.000000, bitrate: 192 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, stereo, fltp, 192 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (mp3 (mp3float) -> pcm_s16le (native))
Press [q] to stop, [?] for help
Output #0, pulse, to 'stream name':
  Metadata:
    icy-name        : My Music on rz4UOJ9JspE
    StreamTitle     : 太阳 - PikA 好声音
    encoder         : Lavf58.20.100
    Stream #0:0: Audio: pcm_s16le, 44100 Hz, stereo, s16, 1411 kb/s
    Metadata:
      encoder         : Lavc58.35.100 pcm_s16le
size=N/A time=00:00:06.47 bitrate=N/A speed=1.25x
```

### CODE REVIEW

File: src/transcode.c [hunk 1]
```diff
@@ -281,6 +287,7 @@ stream_settings_set(struct stream_ctx *s, struct settings_ctx *settings, enum AV
       s->codec->channels       = settings->channels;
       s->codec->sample_fmt     = settings->sample_format;
       s->codec->time_base      = (AVRational){1, settings->sample_rate};
+      s->codec->bit_rate       = settings->bit_rate;
```
Comment (whatdoineed2do) [inline src/transcode.c]: this `AVCodecContext` value was never previously set and was populated by the default value (128000) - setting this to 0 has no impact on day to day transcode ops in testing

**Comment by whatdoineed2do** [comment]
I have also made another upd to the chrome cast as well in a separate branch and change of bitrate appears to work (certainly sounds terrible if drag the bitrste to 6kbps but dont know how to verify higher rates like I could with MP3 stream and ffmpeg)

However I have kept it out of this PR because I'm not sure if the changes have any impact to:

https://github.com/whatdoineed2do/forked-daapd/blob/stream-cast-configurable-bitrate/src/outputs/cast.c:93
`CAST_SAMPLES_PER_PACKET 960`


If you can advise of no impact I can merge both into this PR

File: forked-daapd.conf.in [hunk 2]
```diff
@@ -53,6 +53,9 @@ general {
 	# unusual platform and experience audio drop-outs, you can try changing
 	# this option
 #	high_resolution_clock = yes
+
+	# Set the MP3 streaming bit rate (in kbps), valid options: 64 / 96 / 128 / 192 / 320
```
Comment (ejurgensen) [inline forked-daapd.conf.in]: I think the config option should have its own section. Then we can later add other options. Or you could add say sample_rate right away. Also a better description is needed, otherwise users that don't use /stream.mp3 will be confused by the meaning of this option.

@chme is this change perhaps a candidate for using the new settings feature? If you think so, could you then give some advice on how to do it?

File: src/httpd_streaming.c [hunk 3]
```diff
@@ -52,6 +52,7 @@ extern struct event_base *evbase_httpd;
 #define STREAMING_MP3_SAMPLE_RATE 44100
 #define STREAMING_MP3_BPS         16
 #define STREAMING_MP3_CHANNELS    2
+static int streaming_mp3_bitrate = 192000;
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: I propose you rename the current streaming_quality to streaming_quality_in and make a new streaming_quality_out. The latter is then by default initialised with STREAMING_MP3_ (also a STREAMING_MP3_BIT_RATE), but all the defaults can then be changed via config.

File: src/httpd_streaming.c [hunk 4]
```diff
@@ -86,7 +87,7 @@ static int streaming_meta[2];
 
 #define STREAMING_ICY_METALEN_MAX      4080  // 255*16 incl header/footer (16bytes)
 #define STREAMING_ICY_METATITLELEN_MAX 4064  // STREAMING_ICY_METALEN_MAX -16 (not incl header/footer)
-static const short STREAMING_ICY_METAINT = 8192;
+static const short STREAMING_ICY_METAINT = 16384;
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: Please document why this value was chosen

File: src/httpd_streaming.c [hunk 5]
```diff
@@ -182,7 +183,7 @@ streaming_end(void)
 static void
 streaming_meta_cb(evutil_socket_t fd, short event, void *arg)
 {
-  struct media_quality mp3_quality = { STREAMING_MP3_SAMPLE_RATE, STREAMING_MP3_BPS, STREAMING_MP3_CHANNELS };
+  struct media_quality mp3_quality = { STREAMING_MP3_SAMPLE_RATE, STREAMING_MP3_BPS, STREAMING_MP3_CHANNELS, streaming_mp3_bitrate };
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: See above comment

File: src/httpd_streaming.c [hunk 6]
```diff
@@ -204,6 +205,9 @@ streaming_meta_cb(evutil_socket_t fd, short event, void *arg)
   if (!decode_ctx)
     goto error;
 
+  if (quality.bit_rate)
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: What is the purpose of this?

File: src/outputs/cast.c [hunk 7]
```diff
@@ -439,7 +439,7 @@ static struct cast_session *cast_sessions;
 static struct cast_master_session *cast_master_session;
 //static struct timeval heartbeat_timeout = { HEARTBEAT_TIMEOUT, 0 };
 static struct timeval reply_timeout = { REPLY_TIMEOUT, 0 };
-static struct media_quality cast_quality_default = { CAST_QUALITY_SAMPLE_RATE_DEFAULT, CAST_QUALITY_BITS_PER_SAMPLE_DEFAULT, CAST_QUALITY_CHANNELS_DEFAULT };
+static struct media_quality cast_quality_default = { CAST_QUALITY_SAMPLE_RATE_DEFAULT, CAST_QUALITY_BITS_PER_SAMPLE_DEFAULT, CAST_QUALITY_CHANNELS_DEFAULT, 0 };
```
Comment (ejurgensen) [inline src/outputs/cast.c]: Have you tested that this works? As I understand this means that the Opus encoder will be called with 0 as bit_rate, instead of the current default. How does the encoder deal with that?

Comment (whatdoineed2do) [reply src/outputs/cast.c]: Yes this works - it was tested with a first gen Chromecast (not Chromecast Audio) plugged into the TV that identifies itself like:
```
cast: Device 'cast#00' does not support app id '85CDB22F', trying '0F5096E8' instead
```
I don't know how to determine the output quality .. when we do this with the MP3 stream it appears that ffmpeg goes with 128kbps which is/was the default from the `AVCodecContext`

File: forked-daapd.conf.in (hunk 2 above)
Comment (whatdoineed2do) [reply forked-daapd.conf.in]: @ejurgensen - to allow the users to specify a bit_rate (or perhaps even other params) the server could receive `/stream.mp3?bit_rate=192`.  This would take president over the server config value that we'll hold in `streaming_quality_out.bit_rate` and use this for setting up the mp3 encode for this client.

The web UI would let the user set the required `bit_rate` on the settings page that is then sent to the server.

Does that work for you?

Comment (whatdoineed2do) [reply forked-daapd.conf.in]: Just had another dive into this and it looks like I misunderstood the setup for the transcode and that **theres only one** `transcode_encode_ctx` no matter how many streaming clients we have.

Given this, does it make sense to allow (multiple) streaming clients to potentially request a different set of transcode parameters etc?

Comment (whatdoineed2do) [reply forked-daapd.conf.in]: I've got this `/stream.mp3?bit_rate=xxx` working in another branch with this commit 4c69f5435897d6a6bc84d3eb8f047d198d7f0ca8.  You can start playback and attach a client:
```
ffmpeg -i http://forked-daapd.local:3689/stream.mp3 -f pulse "foo"
```
and you see'll the strream come in with the server defaulted bit rate. In another terminal when you request:
```
ffmpeg -i http://forked-daapd.local:3689/stream.mp3?bit_rate=32 -f pulse "foo"
```
you'll hear that both streams are now at 32kbps (ie awkful so you hear the drop in quality).  With this in place another PR can add the UI settings change (which does not need to be saved in db like the `composer`).

Let me know if you want this work to be part of this PR

Comment (ejurgensen) [reply forked-daapd.conf.in]: I didn't really ask for the above in my review, and I don't agree with it. Like you say a bit_rate parameter entails multiple encodings, and I don't think we should do that. Also I don't think icymetaint should be configurable (it is too low level), and I don't suppose setting bits_per_sample has any meaning when it comes to mp3? And what about the channel param - can ffmpeg actually encode mp3 to more channels than 2?

BTW I have received something like 8 email notifications on this PR alone since my review. That is a bit much. It would be nice if you for instance pushed all changes instead of splitting them.

Comment (whatdoineed2do) [reply forked-daapd.conf.in]: So we drop `bits_per_sample` and `channels` from being configurable and leave them to 16/2 as they are now.

I’m a little hesitant from removing `icy_metaint` as I found that my Roku Soundbridge didn’t like values above 16k - no meta displayed - so I wonder if there are other clients that have such restrictions out there too that may need tweaking this value.  Could we leave this as configurable but remove it from the ‘forked-daapd.conf.in` so it becomes a _hidden_ option like some others I saw in ‘`conffile.c`?

Comment (ejurgensen) [reply forked-daapd.conf.in]: Yes that is ok

**Comment by whatdoineed2do** [comment]
All done and should be ready to merge if you'v got no other concerns - I've also incl the oneline memleak fix that was in a separate PR.

The close callback, `streaming_close_cb()`, cleans up the session but not the `evhhttp_request` object alloc'd from `httpd()-> event_base_dispatch() -> ... -> event_base_loop() -> ... -> evhttp_request_new()`

This leak can be observed by:
* starting the server
* connect a streaming client to the server (`ffmpeg -i http://forked-daapd.local:3689/stream.mp3`)
* kill/disconnect the streaming client
* stop the server
```
==13137== 1,658 (216 direct, 1,442 indirect) bytes in 1 blocks are definitely lost in loss record 369 of 380
==13137==    at 0x4C3114A: calloc (vg_replace_malloc.c:762)
==13137==    by 0x8A433EC: evhttp_request_new (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x8A4352B: ??? (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x8A43740: ??? (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x8A2FC10: ??? (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x8A275B0: ??? (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x8A27D46: event_base_loop (in /usr/lib64/libevent-2.1.so.6.0.2)
==13137==    by 0x425B28: httpd (httpd.c:767)
==13137==    by 0x9EFD593: start_thread (in /usr/lib64/libpthread-2.27.so)
==13137==    by 0xA416E6E: clone (in /usr/lib64/libc-2.27.so)
```
If the server is stopped with active clients the server cleans this up in `session_end()` via `evhttp_send_reply_end(session->req)`

Also, in the future, it may be worth to allow user to set (via undoc cfg option) the `AVCodecContext.compression_level` (defaults to -1, allows range 0-9 with 9 being worst) as per https://github.com/ejurgensen/forked-daapd/issues/804#issuecomment-531318375

**Comment by ejurgensen** [comment]
You're not doing this PR any favor by adding the unrelated memleak change after my review. Please remove it again.

**Comment by whatdoineed2do** [comment]
Removed. The associated memleak PR is also reopened

File: src/conffile.c [hunk 8]
```diff
@@ -60,6 +60,7 @@ static cfg_opt_t sec_general[] =
 #else
     CFG_BOOL("high_resolution_clock", cfg_true, CFGF_NONE),
 #endif
+    CFG_INT("streaming_bitrate", 192, CFGF_NONE),
```
Comment (ejurgensen) [inline src/conffile.c]: Why is this still here?

File: src/httpd_streaming.c [hunk 9]
```diff
@@ -617,6 +625,39 @@ int
 streaming_init(void)
 {
   int ret;
+  cfg_t *cfgsec;
+  int val;
+
+  cfgsec = cfg_getsec(cfg, "streaming");
+
+  val = cfg_getint(cfgsec, "sample_rate");
+  if (val%11025 > 0 && val%12000 > 0 && val%8000 > 0)
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: Add spaces around operators. Also add a comment that this is a checks for libmp3lame_sample_rates. You could also consider dropping the check altogether, because you risk that it gets out of sync with what ffmpeg actually supports. And if the user configures something invalid it will fail anyway with a log message.

File: src/httpd_streaming.c [hunk 10]
```diff
@@ -617,6 +625,39 @@ int
...
+    DPRINTF(E_LOG, L_STREAMING, "non standard streaming sample_rate=%d, defaulting\n", val);
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: Log messages should start with a capital letter. Please fix all of them.

File: src/httpd_streaming.c [hunk 11]
```diff
@@ -86,7 +88,14 @@ static int streaming_meta[2];
...
 #define STREAMING_ICY_METATITLELEN_MAX 4064  // STREAMING_ICY_METALEN_MAX -16 (not incl header/footer)
-static const short STREAMING_ICY_METAINT = 8192;
+
+/* As streaming quality goes up, we send more data to the remote client.  With a
+ * smaller ICY_METAINT value we have to splice metadata more frequently - on 
+ * some devices with small input buffers, a higher quality stream and low 
+ * ICY_METAINT can lead to stuttering as observed on a Roku Soundbridge
+ */
+#define STREAMING_ICY_METAINT_DEFAULT  16384
+static unsigned short STREAMING_ICY_METAINT = STREAMING_ICY_METAINT_DEFAULT;
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: Variables should never be with capital letters!

File: src/httpd_streaming.c [hunk 12]
```diff
@@ -617,6 +625,39 @@ int
...
+      DPRINTF(E_WARN, L_STREAMING, "streaming bit_rate=%d not accepted, defaulting\n", val);
+  }
+  DPRINTF(E_INFO, L_STREAMING, "streaming quality: %d/%d/%d @ %dkbps\n", streaming_quality_out.sample_rate, streaming_quality_out.bits_per_sample, streaming_quality_out.channels, streaming_quality_out.bit_rate/1000);
+
+  val = cfg_getint(cfgsec, "icy_metaint");
+  // Too low a value forces server to send more meta than data
+  if (val >= 4096 && val <= 131072)
+    STREAMING_ICY_METAINT = val;
+  else
+    DPRINTF(E_INFO, L_STREAMING, "icy_metaint=%d not accepted, defaulting to %d\n", val, STREAMING_ICY_METAINT);
```
Comment (ejurgensen) [inline src/httpd_streaming.c]: Improve this message so it tells the user what is wrong -> ie how to correct

File: src/transcode.c [hunk 13]
```diff
@@ -731,7 +738,7 @@ open_decoder(unsigned int *stream_index, struct decode_ctx *ctx, enum AVMediaTyp
   if ((*stream_index < 0) || (!decoder))
     {
       if (!ctx->settings.silent)
-	DPRINTF(E_LOG, L_XCODE, "No stream data or decoder for stream #%d\n", *stream_index);
+       DPRINTF(E_LOG, L_XCODE, "No stream data or decoder for stream #%d\n", *stream_index);
```
Comment (ejurgensen) [inline src/transcode.c]: Why is this changed?

File: src/httpd_streaming.c (hunk 9 above)
Comment (whatdoineed2do) [reply src/httpd_streaming.c]: I think its cleaner to provide correct values for setting up the ffmpeg/transcode stuf - if we let the values pass through ffmpeg/transcode will fail setting up and disable streaming requiring restart which people won't notice until a streaming client attaches  With this at least we have a working streaming even if its not with the params the user (thought) they'd config'd.

**Comment by ejurgensen** [comment]
Merged, thanks
